### PR TITLE
Ability to copy and save others playlists

### LINKED ISSUE #1879: Ability to create (copy) playlist from Favorites / Watch later (or anyone's playlist)
@tzarebczan commented on [Thu Jun 10 2021](https://github.com/lbryio/lbry-desktop/issues/6210)

We discussed this as a nice to have in the future - I can click an option to start a new playlist from my watch later/favorite/own or others' playlist.

### PR DESCRIPTION
**Pull request by saltrafael**
## Fixes

Issue Number: closes #1879

- Copy: makes a new private playlist from another existing playlist
- Save: Adds someone's public playlist to a saved list and displays them on your playlists page. Basically allows to "follow" future updates and always have the most recent version of the playlist, but can't edit or publish it (unless copied).

🟢  https://salt.odysee.com

### CODE REVIEW

File: ui/redux/reducers/collections.js [hunk 1]
```diff
@@ -189,12 +200,13 @@ const collectionsReducer = handleActions(
...
+
       return {
         ...state,
         edited: editedCollections || state.edited,
         unpublished: unpublishedCollections || state.unpublished,
         builtin: builtinCollections || state.builtin,
-        saved: savedCollections || state.saved,
+        saved: Object.assign(state.saved, savedCollections),
       };
     },
```
Comment (infinite-persistence) [inline ui/redux/reducers/collections.js:211]: :small_red_triangle: Changes must be immutable

The use of `Object.assign` on an array ended up altering the first argument `state.saved` directly.

image

Feels like the original code is correct (use `savedCollections`, or fallback to original `state.saved` if no changes), unless I missed something.

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: ui/redux/actions/collections.js [hunk 2]
```diff
@@ -3,36 +3,67 @@ import * as ACTIONS from 'constants/action_types';
...
 
+  const id = uuid(); // start with a uuid, this becomes a claimId after publish
   if (cb) cb(id);
 
+  if (sourceId) {
+    const state = getState();
+    const sourceCollectionItems = selectUrlsForCollectionId(state, sourceId);
+    const sourceCollection = selectCollectionForId(state, sourceId);
+    const sourceCollectionClaim = selectClaimForId(state, sourceId);
+    const sourcedescription =
```
Comment (infinite-persistence) [inline ui/redux/actions/collections.js:43]: `sourceDescription`

File: ui/component/collectionItemsList/view.jsx [hunk 3]
```diff
@@ -36,13 +36,10 @@ const CollectionItemsList = (props: Props) => {
     ...claimListProps
   } = props;
```
Comment (infinite-persistence) [inline ui/component/collectionItemsList/view.jsx:37]: In the props definittion, `doFetchItemsInCollection: ({}, ?() => void) => void,` needs an update since the callback was removed.

File: ui/page/playlists/internal/collectionsListMine/view.jsx [hunk 4]
```diff
@@ -56,20 +64,26 @@ export default function CollectionsListMine(props: Props) {
...
           return unpublishedCollectionsList;
         case COLS.LIST_TYPE.PUBLIC:
           return publishedList;
+        case COLS.LIST_TYPE.EDITED:
+          return editedList;
+        case COLS.LIST_TYPE.SAVED:
+          return savedList;
       }
-    }, [filterType, publishedList, unpublishedCollectionsList]) || [];
+    }, [editedList, filterType, publishedList, savedList, unpublishedCollectionsList]) || [];
```
Comment (infinite-persistence) [inline ui/page/playlists/internal/collectionsListMine/view.jsx:86]: :arrow_down: Just a discussion

- Considering there could be users with large number of lists, I wonder if concat is faster.
  -  https://stackoverflow.com/a/57191593/977819. 
  - https://www.measurethat.net/Benchmarks/Show/579/1/arrayprototypeconcat-vs-spread-operator
- The `switch :: default` could return the `[ ]` rather than using an explicit fallback outside of `memo` (minor).

File: ui/page/collection/internal/collectionHeader/internal/collectionSubtitle/view.jsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// @flow
+import React from 'react';
+import ClaimDescription from 'component/claimDescription';
+import ClaimAuthor from 'component/claimAuthor';
+import CollectionPrivateIcon from 'component/common/collection-private-icon';
+import Skeleton from '@mui/material/Skeleton';
+
+type Props = {
+  // -- redux --
+  uri?: string,
```
Comment (infinite-persistence) [inline ui/page/collection/internal/collectionHeader/internal/collectionSubtitle/view.jsx:10]: `uri` is client side, not redux (minor cleanup)

File: ui/page/collection/internal/collectionHeader/internal/collectionSubtitle/index.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+import { selectClaimForId } from 'redux/selectors/claims';
+import {
+  selectCollectionDescriptionForId,
+  selectCountForCollectionId,
+  selectCollectionHasEditsForId,
+  selectMyPublishedCollectionCountForId,
+} from 'redux/selectors/collections';
+
+import CollectionHeader from './view';
```
Comment (infinite-persistence) [inline ui/page/collection/internal/collectionHeader/internal/collectionSubtitle/index.js:11]: Could clean up a bunch of `CollectionHeader` in the index files, but minor (there are an extension of `CollectionHeader` after all)

**Review by infinite-persistence — COMMENTED** [review]
Didn't re-test to see if lists are resolving individually or not, but should be fine. I think the removal of `didTryResolve` shouldn't impact.

File: flow-typed/Claim.js [hunk 7]
```diff
@@ -82,7 +82,7 @@ declare type ChannelMetadata = GenericMetadata & {
```
Comment (infinite-persistence) [inline flow-typed/Claim.js:82]: :small_red_triangle: I had a public collection (I think it was public) with only 1 item. I then edited it and removed that 1 entry. The GUI goes into a bad state. When I return to `/$/playlists`, I tile is in a bad state as well, so I can't undo changes.

File: ui/page/playlists/internal/collectionsListMine/view.jsx (hunk 4 above)
Comment (saltrafael) [reply ui/page/playlists/internal/collectionsListMine/view.jsx]: Good points 👍🏻

File: ui/component/formNewCollection/view.jsx [hunk 8]
```diff
@@ -10,56 +10,65 @@ import { formatLbryUrlForWeb } from 'util/url';
...
     location: { search },
   } = useHistory();
 
   const buttonref: ElementRef<any> = React.useRef();
-  const newCollectionName = React.useRef('');
 
-  const [disabled, setDisabled] = React.useState(true);
+  const [newCollectionName, setCollectionName] = React.useState(
+    __('%copied_playlist_name% (copy)', { copied_playlist_name: sourceCollectionName }) || ''
+  );
```
Comment (infinite-persistence) [inline ui/component/formNewCollection/view.jsx:39]: The fallback `|| ''` will never hit, since the front part will always be a valid string, right?

File: ui/redux/selectors/collections.js [hunk 9]
```diff
@@ -40,6 +40,26 @@ export const selectUnpublishedCollectionsList = createSelector(
...
 
+export const selectCollectionSavedForId = (state: State, id: string) => {
+  const savedIds = selectSavedCollectionIds(state);
+  return savedIds.includes(id);
+};
+
+export const selectSavedCollections = createSelector(
+  selectSavedCollectionIds,
+  (state) => (id) => selectCollectionForId(state, id),
+  (savedIds, collectionForId) => {
```
Comment (infinite-persistence) [inline ui/redux/selectors/collections.js:51]: I haven't loaded to fully try it, but I think this will not memoize correctly because the second selector, `(state) => (id) => selectCollectionForId(state, id),`, will always return a new inline function, so `selectSavedCollections` will always re-evaluate.

But I also see the problem .. you would then need to bring along every selector encapsulated by `selectCollectionForId` here..

```
  selectSavedCollectionIds,
  // plus
  selectBuiltinCollections,
  selectResolvedCollections,
  selectMyUnpublishedCollections,
  selectMyEditedCollections,
  selectPendingCollections,
  selectCurrentQueueList,
  (... ) => {}
```

If cannot find a simpler solution, I guess not cache it for now (use simple function), since this won't be caching correctly.

File: ui/component/formNewCollection/view.jsx (hunk 8 above)
Comment (saltrafael) [reply ui/component/formNewCollection/view.jsx:39]: Yep you are right

File: ui/redux/selectors/collections.js (hunk 9 above)
Comment (saltrafael) [reply ui/redux/selectors/collections.js:51]: do you think this is any different? https://github.com/OdyseeTeam/odysee-frontend/commit/d32e0a84e36fbd7290a2571767a63e35cce6df59

File: ui/component/formNewCollection/view.jsx [hunk 10]
```diff
@@ -10,56 +10,63 @@ import { formatLbryUrlForWeb } from 'util/url';
...
   const {
     push,
     location: { search },
   } = useHistory();
 
   const buttonref: ElementRef<any> = React.useRef();
-  const newCollectionName = React.useRef('');
 
-  const [disabled, setDisabled] = React.useState(true);
+  const [newCollectionName, setCollectionName] = React.useState(sourceCollectionName || '');
```
[the remark this answers is not in the document]
Comment (saltrafael) [reply ui/component/formNewCollection/view.jsx]: I think `(copy)` alone is enough? The numbers don't feel necessary, you could still sort by creation date if needed. There will be multiple playlists with the same name but it's expected since they are all the same copy?

File: ui/redux/selectors/collections.js (hunk 9 above)
Comment (infinite-persistence) [reply ui/redux/selectors/collections.js]: Yeah, saving into redux is a good alternative if the selector got too complex.

- :small_blue_diamond: `savedCollections: { source: 'collections', property: 'savedIds' },` this got very confusing because there is already a state in redux called `savedCollections`.  It's also confusing when viewing the wallet because it's not a collection map like the rest, but an array of collection IDs.
- :grey_question: Can you summarize the purpose of `withEdits`?
  - `collectionsWithEdits` and `editedCollections` mean the same thing at first glance of the wallet, so that's confusing.
  - For the most part, they seem to duplicates, even for the timestamps.
  - When I clicked "clear edits", that should just restore the collection back into the published state and remove it from `editedCollections`, right? But the wallet still contains the `editedCollections` entry.  What does this scenario represent?
    - image

File: flow-typed/Claim.js (hunk 7 above)
Comment (saltrafael) [reply flow-typed/Claim.js:82]: Does this still happen?

File: ui/redux/selectors/collections.js (hunk 9 above)
Comment (saltrafael) [reply ui/redux/selectors/collections.js]: - Improved it a bit, could've used the `resolved` object all this time for saved. The saved collections are stored as an array of ids since they are public playlists, they need to be resolved to store the current state (they can be updated, deleted, etc)
- `withEdits` now `update` the only purpose is to delete it from `edited` but still keep track of when was last updated, since `edited` only applies to public collections, it's safe to delete the `edited` entry when edits are cleared since it will then select the `resolved` entry (without edits)

File: flow-typed/Claim.js (hunk 7 above)
Comment (infinite-persistence) [reply flow-typed/Claim.js:82]: Yes.  In this video, it also highlights some oddities when Watch Later is empty.

https://user-images.githubusercontent.com/64950861/181424040-5b3bee9a-87ad-40ff-ba34-bd6cdf8e2d3f.mp4

File: ui/redux/selectors/collections.js (hunk 9 above)
Comment (infinite-persistence) [reply ui/redux/selectors/collections.js]: It much cleaner now, thanks.

- [ ] :small_blue_diamond: `updatedCollections` is still writing the entire collection.
  - In `SharedData`, it is defined as `updatedCollections: UpdatedCollection`.  I think that's good to not bloat the wallet.
  - In the wallet, it is written as a full `Collection`.  I think need to transform the data in `sharedStateFilters`.
- [ ] :grey_question: If I cleared an edit, and later edit again + really publish it, does it make sense for the `updatedCollections` entry to still be there?  It should probably just use the claim's timestamp at that point.
  - Back in [previous comment](https://github.com/OdyseeTeam/odysee-frontend/pull/1877#discussion_r930599284), I was confused because I expected "Last Updated At" to just revert to the claim's timestamp when edits are cleared, so `withEdits | updated` felt unnecessary -- simpler model.  But I can also see the benefit of listing when it was last touched, even if  reverted.

File: ui/redux/selectors/collections.js [hunk 11]
```diff
@@ -40,6 +41,17 @@ export const selectUnpublishedCollectionsList = createSelector(
...
 
+export const selectCollectionSavedForId = (state: State, id: string) => {
+  const savedIds = selectSavedCollectionIds(state);
+  return savedIds.includes(id);
+};
+
+export const selectSavedCollections = createSelector(
+  selectResolvedCollections,
+  selectSavedCollectionIds,
+  (resolved, savedIds) => Object.fromEntries(Object.entries(resolved).filter(([key, val]) => savedIds.includes(key)))
```
Comment (infinite-persistence) [inline ui/redux/selectors/collections.js:52]: I wonder if we still need to use the `fromEntries` polyfill like the rest in this file.
caniuse seems to indicate most browsers support it since 2019, but there must be a reason a polyfill was used (user complaint?)

Comment (infinite-persistence) [reply ui/redux/selectors/collections.js:52]: https://odysee-workspace.slack.com/archives/C02PTRGFR70/p1628123205030100

File: ui/redux/selectors/collections.js (hunk 9 above)
Comment (saltrafael) [reply ui/redux/selectors/collections.js]: Fixed `updatedCollections` and looks like removing edited was a thing so now it should delete updated instead (should be even better than before now since you keep your edited version, let's say you used LBRY desktop to publish the collection but forgot you had added some other items on odysee, the edits shouldn't be cleared by a most recent version, but I didn't test this)

File: flow-typed/Claim.js (hunk 7 above)
Comment (saltrafael) [reply flow-typed/Claim.js:82]: Now it should be fixed?

**Comment by infinite-persistence** [comment]
[b96b0c9 ](https://github.com/OdyseeTeam/odysee-frontend/commit/b96b0c9730e93db5f2261366169bd1cb0c3edd15)to [512e10d ](https://github.com/OdyseeTeam/odysee-frontend/commit/512e10dd272f6d1a1a86d35db346ca3baf7ac394)
- Squash to minimal; prune discarded ideas. Backup available in `backup/` branch.
